# Buelengde

Teorien om integrasjon kan også brukes til å finne lengden av en graf eller parametrisert kurve, noe som kalles for *buelengde*.

## Lengden av en graf

Tenk deg at vi har en funksjon $f(x)$ og vi ønsker å regne ut lengden til grafen over et intervall $[a, b]$. Måten vi kan tilnærme lengden dens på, er å dele opp intervallet i $n$ delintervaller slik at vi får $n + 1$ punkter $x_0, x_1, \ldots, x_{n-1}$ som er jevnt fordelt i intervallet. For hvert punkt $x_i$, kan vi regne ut tilsvarende funksjonsverdier $y_i = f(x_i)$ slik at vi får $n + 1$ $y$-verdier $y_0, y_1, \ldots, y_n$. Med denne oppbygningen kan vi skrive ned resultatet som vi baserer oss på:

```{prf:theorem} Lengde av graf
:label: lengde_av_graf

La $f(x)$ være en funksjon definert på et intervall $[a, b]$. La $x_0, x_1, \ldots, x_n$ være $n + 1$ punkter som er jevnt fordelt i intervallet. Da kan vi lage oss $n$ rettvinklede trekanter med kateter

\begin{align*}
\Delta x & = (b - a) / n, \\
\\
\Delta y_i & = f(x_i + \Delta x) - f(x_i),
\end{align*}

der $x_i = a + i\Delta x$ for $i = 0, 1, 2, \ldots, n$. 

Trekantene har da hypotenuser med lengde

$$
\Delta s_i = \sqrt{\Delta x^2 + \Delta y_i^2},
$$

Lengden av grafen til $f(x)$ over intervallet $[a, b]$ blir da tilnærmet gitt ved

$$
L \approx \sum_{i=0}^{n - 1} \Delta s_i = \Delta s_0 + \Delta s_1 + \ldots + \Delta s_{n - 1},
$$
```

Størrelsen $\Delta s$ kalles for et **linjeelement** og er et konsept som dukker opp i fysikken i mange sammenhenger.


I {numref}`lengde_graf` ser du en visualisering av algoritmen.


```{figure} ../../../codes/math/arc_length/arc_length_illustration.pdf
:name: lengde_graf

Visualisering av hvordan man kan estimere lengden til grafen til $f(x) = x^2$ mellom $x = 1$ og $x = 4$ med lengden av hypotenusen til trekantene spent ut av $\Delta x_i$ og $\Delta y_i$ som kateter. Her har vi brukt 3 segmenter. 
```


### Eksempel 1

La oss bruke algoritmen over til å regne ut lengden av grafen

$$
f(x) = x^2
$$

på intervallet $[1, 4]$ slik som i {numref}`lengde_graf`. Vi skal bruke $n = 100$ punkter i stedet for $n = 4$ som i figuren. 

Vi kan kode dette slik:

In [55]:
import numpy as np

def f(x):
    return x**2

# Definerer intervallet [a, b] og antall punkter n
a = 1
b = 4
n = 10
dx = (b - a) / n # Steglengde Δx

lengde = 0
for i in range(n):
    x = a + i * dx
    dy = f(x + dx) - f(x) # Δy
    
    ds = np.sqrt(dx**2 + dy**2) # Δs

    lengde += ds # Legger til Δs til den totale lengden

print(f"{lengde = :.2f}")

lengde = 15.34


### Eksempel 2

La oss bruke algoritmen til å regne ut lengden av grafen

$$
g(x) = \sqrt{1 + \cos^2(x)}, \quad x \in [0, 2\pi].
$$

La oss bruke $n = 1001$ punkter (1000 delintervaller). En kode som løser dette kan se slik ut:

In [57]:
import numpy as np

def f(x):
    return np.sqrt(1 + np.cos(x)**2)

# Definerer intervallet [a, b] og antall punkter n
a = 0
b = 2 * np.pi
n = 1000
dx = (b - a) / n # Δx

lengde = 0
for i in range(n):
    x = a + i * dx
    dy = f(x + dx) - f(x) # Δy
    ds = np.sqrt(dx**2 + dy**2) # Δs

    lengde += ds

print(f"{lengde = :.2f}")

lengde = 6.54


## Lengden av en parametrisert kurve

Vi skal se på hvordan vi kan regne ut lengden til parametriserte kurver i både to- og tre dimensjoner. Det viser seg at teorien er akkurat den samme for de to tilfellene, men at det i tre dimensjoner dukker opp ett ekstra ledd for $z$-retningen noen av likningene.
Uansett skal vi tenke oss at vi har parametriserte vektorer på formen

\begin{align*}
\vec{r}(t) & = \left[x(t), y(t)\right] \quad\quad\quad \text{(2 dimensjoner)}, \\
\\
\vec{r}(t) & = \left[x(t), y(t), z(t)\right] \quad \text{(3 dimensjoner)}. \\
\end{align*}

### Algoritme: Summere Linjeelementer $\Delta s$

Denne metoden er i stor grad uendret. Vi regner fortsatt ut lengden av hypotenusen i $n$ trekanter. Den store forskjellen er vi nå ikke direkte setter verdiene til $x$, men lar heller parameteren $t$ bestemme hva verdien til $x$ og $y$ er. I tillegg vil den bestemme hva verdien til $z$ er når vi jobber i tre dimensjoner. Følgende resultater gir oss verktøyene vi trenger for å regne ut lengden av en parametrisert kurve:

```{prf:theorem} Lengde av kurve (2 dimensjoner)
:label: lengde_kurve_2d

La $\vec{r}(t) = \left[x(t), y(t)\right]$ være en parametrisert kurve i to dimensjoner og $t \in [a, b]$ er en parameter. Vi introduserer $n + 1$ jevnt fordelte punkter i parameterrommet $t_0, t_1, t_2, \ldots, t_n$. Da er linjeelementene $\Delta s_i$ gitt ved

$$
\Delta s_i = \sqrt{\Delta x_i^2 + \Delta y_i^2},
$$

der 

\begin{align*}
\Delta x_i & = x(t_i + \Delta t) - x(t_i), \\
\Delta y_i & = y(t_i + \Delta t) - y(t_i),
\end{align*}

der $t_i = a + i\Delta t$ for $i = 0, 1, 2, \ldots, n$, og $\Delta t = (b - a) / n$. Lengden av kurven er da tilnærmet gitt ved

$$
L \approx \sum_{i=0}^{n-1} \Delta s_i = \Delta s_0 + \Delta s_1 + \ldots + \Delta s_{n-1}.
$$
```

```{prf:theorem} Lengde av kurve (3 dimensjoner)
:label: lengde_kurve_3d

La $\vec{r}(t) = \left[x(t), y(t), z(t)\right]$ være en parametrisert kurve i tre dimensjoner og $t \in [a, b]$ er en parameter. 
La $t_0, t_1, \ldots, t_n$ være $n + 1$ punkter i parameterrommet som er jevnt fordelt.
Da er linjeelementene $\Delta s_i$ gitt ved

$$
\Delta s_i = \sqrt{\Delta x_i^2 + \Delta y_i^2},
$$

der 

\begin{align*}
\Delta x_i & = x(t_i + \Delta t) - x(t_i), \\
\Delta y_i & = y(t_i + \Delta t) - y(t_i), \\
\Delta z_i & = z(t_i + \Delta t) - z(t_i),
\end{align*}

der $t_i = a + i\Delta t$ for $i = 0, 1, 2, \ldots, n$ og $\Delta t = (b - a) / n$. Lengden av kurven er da tilnærmet gitt ved

$$
L \approx \sum_{i=0}^{n-1} \Delta s_i = \Delta s_0 + \Delta s_1 + \ldots + \Delta s_{n-1}.
$$
```

```{admonition} Sammenheng mellom lengde av graf og lengde av kurve
:class: tip, dropdown

Det viser seg at vi i starten bare så på et spesialtilfelle av resultatet i to dimensjoner. Vi kan faktisk bruke samme metode for å regne ut lengden av en graf ved å tenke på funksjonen vår som en parametrisering på formen

$$
\vec{r}(x) = \left[x, f(x)\right],
$$

der $x \in [a, b]$ er parameteren. 
```

### Eksempel 3
En arkimedisk spiral (oppkalt etter den greske matematikeren Arkimedes fra antikkens Hellas) er en kurve som er vist i {numref}`archimedean_spiral`. 

```{figure} ./figurer/archimedean_spiral.pdf
:name: archimedean_spiral

En arkimedisk spiral.
```

Den arkimediske spiralen i {numref}`archimedean_spiral` er gitt ved parametriseringen

$$
\vec{r}(t) = \left[t\cos(t), t\sin(t)\right], \quad t \in [0, 4\pi].
$$

Vi ønsker å finne lengden av en bue av spiralen fra $t=0$ til $t=4\pi$. Vi bruker metoden ved å regne ut linjeelementer $\Delta s$ som en tilnærming til lengden av kurven. 

En kode som løser dette kan se slik ut:

In [58]:
import numpy as np

def x(t):
    return t * np.cos(t)


def y(t):
    return t * np.sin(t)


a = 0
b = 4 * np.pi 
n = 10
dt = (b - a) / n # Δt

lengde = 0
for i in range(n): # går til n - 1 fordi vi trenger t + dt til å være siste tidspunkt
    t = a + i * dt
    dx = x(t + dt) - x(t) # Δx
    dy = y(t + dt) - y(t) # Δy
    ds = np.sqrt(dx**2 + dy**2) # Δs

    lengde += ds


print(f"{lengde = :.2f}")

t + dt = 12.566370614359172
b = 12.566370614359172
lengde = 75.16


### Eksempel 4

En *heliks* er en kurve som beveger seg langs randen av en sylinder. Et eksempel på der denne kurven dukker opp i naturen er når ladde partikler fra solstormer følger jordens magnetfelt inn mot polene! I {numref}`helix` ser vi en skisse av en heliks.

```{figure} ./figurer/helix.pdf
:name: helix

En tre-dimensjonal heliks. Pilhoder er tegnet inn for å indikere retningen til heliksen i rommet.
```

Heliksen i {numref}`helix` er beskrevet av parametriseringen

$$
\vec{r}(t) = \left[\cos(t), \sin(t), t\right], \quad t \in [0, 10\pi].
$$

Vi kan bruke metoden med linjeelementer til å finne lengden av heliksen på følgende måte:

In [59]:
import numpy as np

def x(t):
    return np.cos(t)


def y(t):
    return np.sin(t)


def z(t):
    return t


a = 0 # Startverdien til t
b = 10 * np.pi # Sluttverdien til t
n = 100_000 # antall linjeelementer
dt = (b - a) / n # Δt

lengde = 0
for i in range(n):
    t = a + i * dt # Regn ut t-verdien

    # Regn ut katetene i 3-dimensjonal trekant
    dx = x(t + dt) - x(t) # Δx
    dy = y(t + dt) - y(t) # Δy
    dz = z(t + dt) - z(t) # Δz

    # Regn ut hypotenusen
    ds = np.sqrt(dx**2 + dy**2 + dz**2) # Δs

    # Legg til bidraget til lengden
    lengde += ds


print(f"{lengde = :.2f}")

lengde = 44.43


## Oppgaver

### Oppgave 1

Finn en tilnærming til lengden av grafen til funksjonen 

$$
f(x) = e^{-x}x^2 \quad \text{for} \quad x \in [0, 2],
$$

Bruk $n = 1000$ punkter.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

def f(x):
    return NotImplemented

# Definer intervallet, antall punkter og steglengde
a = NotImplemented
b = NotImplemented
n = NotImplemented
dx = NotImplemented # Δx

lengde = 0
for i in range(n):
    x = NotImplemented # x_i
    dy = NotImplemented # Δy
    ds = NotImplemented # Δs

    lengde = NotImplemented

print(f"{lengde = :.2f}")

````{dropdown} Løsningsforslag
```python
import numpy as np

def f(x):
    return x**2 * np.exp(-x)


a = 0
b = 2
n = 1000
dx = (b - a) / n # Δx

lengde = 0
for i in range(n):
    x = a + i * dx
    dy = f(x + dx) - f(x) # Δy
    ds = np.sqrt(dx**2 + dy**2) # Δs

    lengde += ds

print(f"{lengde = :.2f}")
```
som gir utskriften
```console
lengde = 2.09
```
````

### Oppgave 2

Bruk metoden med linjeelementer $\Delta s_i$ for å finne en tilnærming til verdien til $\pi$ ved å beregne lengden til grafen til funksjonen

$$
g(x) = \sqrt{1 - x^2}, \quad x \in [-1, 1].
$$

Funksjonen beskriver øvre halvdel av enhetssirkelen. Bruk 1 million punkter.


*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

def g(x):
    return NotImplemented

a = NotImplemented
b = NotImplemented
n = NotImplemented
h = NotImplemented

pi = 0
for i in range(n):
    dx = NotImplemented # Δx
    dy = NotImplemented # Δy
    ds = NotImplemented # Δs

    pi = NotImplemented

print(f"{pi = :.7f}")

````{dropdown} Løsningsforslag
```python
import numpy as np

def g(x):
    return np.sqrt(1 - x**2)

a = -1
b = 1
n = 1_000_000
h = (b - a) / n

pi = 0
for i in range(n):
    x = a + i * h # x_i
    dx = h # Δx
    dy = g(x + h) - g(x) # Δy
    ds = np.sqrt(dx**2 + dy**2) # Δs

    pi += ds

print(f"{pi = :.7f}")
```
som gir utskriften
```console 
pi = 3.1415926
```
````

### Oppgave 3

Høydeprofilen til en løype i et VM-løp i langrenn er beskrevet av funksjonen

$$
f(x) = 2.5x^3 - 34.7x^2 + 120x + 530, \quad x \in [0, 10],
$$


der $x$ er målt i antall kilometer i luftlinje og $f$ beskriver høyden i meter over havet. I {numref}`ski_race`` ser du en skisse av høydeprofilen.

```{figure} ./figurer/ski_race.pdf
:name: ski_race

Skisse av høydeprofilen til en VM-løypa. Løypa er 10 km lang i luftlinje. $x$-aksen er målt i kilometer og $y$-aksen er målt i meter over havet.
```


Estimer hvor lang løypa er med metodene beskrevet i {prf:ref}`lengde_av_graf`. 

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`. Husk at $x$ og $f(x)$ har forskjellige enheter når du regner ut lengden.*

In [63]:
import numpy as np

def f(x):
    return 2.5 * x**3 - 34.7 * x**2 + 120 * x + 530

a = 0
b = 10
n = 10_000
dx = (b - a) / n # Δx

lengde = 0
for i in range(n):
    x = a + i * dx # x_i
    dy = f(x + dx) - f(x) # Δy
    ds = np.sqrt((dx * 1000)**2 + dy**2) # Δs

    lengde += ds

print(f"{lengde = :.0f} meter.")

lengde = 10020 meter.


In [ ]:
import numpy as np

def f(x):
    return NotImplemented

# Definer intervallet, antall punkter og steglengden.
a = NotImplemented
b = NotImplemented
n = NotImplemented
dx = NotImplemented # Δx

lengde = 0
for i in range(n):
    x = NotImplemented # x_i
    dy = NotImplemented # Δy

    # Husk at dx er målt i kilometer
    # Mens dy er målt i meter
    # Du må konvertere dx til meter! 
    ds = NotImplemented # Δs

    lengde = NotImplemented

print(f"{lengde = :.0f} meter.")

````{dropdown} Løsningsforslag
```python
import numpy as np

def f(x):
    return 2.5 * x**3 - 34.7 * x**2 + 120 * x + 530

a = 0
b = 10
n = 10_000
dx = (b - a) / n # Δx

lengde = 0
for i in range(n):
    x = a + i * dx # x_i
    dy = f(x + dx) - f(x) # Δy
    ds = np.sqrt((dx * 1000)**2 + dy**2) # Δs

    lengde += ds

print(f"{lengde = :.0f} meter.")
```
som gir utskriften
```console
lengde = 10020 meter.
```
Altså er lengden av grafen $L \approx 10020 \, \text{m}$.
````

### Oppgave 4

En *sykloide* er kurven vi får dersom vi fester en penn til et hjul som ruller langs en rett linje.

```{figure} ./animasjoner/Cycloid.gif
:name: cycloid

Animasjon av hvordan kurven til en sykloide dannes. Den rød kurven er sykloiden. Laget av Zorgit, CC BY-SA 3.0, [Wikimedia Commons](https://commons.wikimedia.org/w/index.php?curid=4552689).
```

Kurven med et hjul som har radius lik 1, og som ruller langs den positive $x$-aksen, kan beskrives med parametriseringen

$$
\vec{r}(t) = \left[(t - \sin(t)), (1 - \cos(t))\right],  \qquad t \in [0, 6\pi],
$$

der hjulet gjør tre fulle rotasjoner (én rotasjon per $2\pi$). 


Estimer lengden til sykloiden.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

def x(t):
    return NotImplemented

def y(t):
    return NotImplemented

a = NotImplemented # Startverdi på intervallet
b = NotImplemented # Sluttverdi på intervallet
n = NotImplemented # Antall intervaller
dt = NotImplemented # Δt

lengde = 0 
for i in range(n):
    t = NotImplemented # t-verdien

    dx = NotImplemented # Δx
    dy = NotImplemented # Δy
    
    ds = NotImplemented # Δs

    lengde += NotImplemented

print(f"{lengde = :.2f}")

````{dropdown} Løsningsforslag
```python
import numpy as np

def x(t):
    return t - np.sin(t)

def y(t):
    return 1 - np.cos(t)

a = 0
b = 6 * np.pi
n = 100_000
dt = (b - a) / n

lengde = 0 
for i in range(n):
    t = a + i * dt 

    dx = x(t + dt) - x(t) # Δx
    dy = y(t + dt) - y(t) # Δy
    
    ds = np.sqrt(dx**2 + dy**2) # Δs

    lengde += ds

print(f"{lengde = :.2f}")
```
som gir utskriften
```console
lengde = 24.00
```
````

### Oppgave 5

En logaritmisk spiral er vist i {numref}`logarithmic_spiral`

```{figure} ./figurer/logarithmic_spiral.pdf
:name: logarithmic_spiral

En logaritmisk spiral.
```

En parametrisering av denne kurven er gitt ved

$$
\vec{r}(t) = \left[e^t\cos(4t), e^t\sin(4t) \right], \quad t \in [0, 2\pi].
$$

Finn et estimat av lengden av kurven. 

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

def x(t):
    return NotImplemented

def y(t):
    return NotImplemented

a = NotImplemented
b = NotImplemented
n = NotImplemented
dt = NotImplemented # steglengde Δt

lengde = 0
for i in range(n):
    t = NotImplemented

    dx = NotImplemented # Δx
    dy = NotImplemented # Δy

    ds = NotImplemented # Δs

    lengde = NotImplemented

print(f"{lengde = :.2f}")

````{dropdown} Løsningsforslag
```python
import numpy as np

def x(t):
    return np.exp(t) * np.cos(4 * t)

def y(t):
    return np.exp(t) * np.sin(4 * t)

a = 0
b = 2 * np.pi
n = 100_000
dt = (b - a) / n 

lengde = 0
for i in range(n):
    t = a + i * dt

    dx = x(t + dt) - x(t)
    dy = y(t + dt) - y(t)

    ds = np.sqrt(dx**2 + dy**2)

    lengde += ds

print(f"{lengde = :.2f}")
```
som gir utskriften
```console
lengde = 2203.77
```
````

### Oppgave 6

*Bernoullis Lemniskate* er en lukket sløyfe som kan sies å likne på et liggende åttetall eller på et $\infty$-symbol. Kurven vises i {numref}`lemniscate_of_bernoulli`


```{figure} ./figurer/lemniscate_of_bernoulli.pdf
:name: lemniscate_of_bernoulli
:align: center
Figuren viser Bernoullis lemniskate.
```

Kurven kan parametriseres med

$$
\vec{r}(t) = \left[\frac{\cos(t)}{1 + \sin^2(t)}, \frac{\cos(t)\sin(t)}{1 + \sin^2(t)}\right], \qquad t \in [0, 2\pi].
$$

Regn ut en tilnærming lengden av kurven.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

# x-komponenten til parametriseringen
def x(t):
    return NotImplemented


# y-komponenten til parametriseringen
def y(t):
    return NotImplemented

a = NotImplemented  # Startverdi for t
b = NotImplemented # sluttverdi for t
n = NotImplemented # antall intervaller
dt = NotImplemented # steglengde Δt

lengde = 0
for i in range(n):
    t = NotImplemented # Regn ut t-verdien

    dx = NotImplemented # Regn ut Δx
    dy = NotImplemented # Regn ut Δy
    ds = NotImplemented # Regn ut Δs
    
    lengde += NotImplemented # legg til bidraget 

print(f"{lengde = :.2f}")

````{dropdown} Løsningsforslag
```python
import numpy as np

# x-komponenten til parametriseringen
def x(t):
    return np.cos(t) / (1 + np.sin(t)**2)


# y-komponenten til parametriseringen
def y(t):
    return np.cos(t) * np.sin(t) / (1 + np.sin(t)**2)

a = 0
b = 2 * np.pi
n = 100_000
dt = (b - a) / n

lengde = 0
for i in range(n):
    t = a + i * dt

    dx = x(t + dt) - x(t)
    dy = y(t + dt) - y(t)
    ds = np.sqrt(dx**2 + dy**2)

    lengde += ds

print(f"{lengde = :.2f}")
```
som gir utskriften
```console
lengde = 5.24 
```
````

### Oppgave 7

En *astroide* i matematikken er en kurve som genereres ved at en sirkel ruller på innsiden av en annen sirkel og gjennomfører en full rotasjon per kvadrant. I {numref}`astroid` kan du se hvordan kurven lages og ser ut.

```{figure} ./animasjoner/astroid.gif
:name: astroid

En astroide. Den røde kurven er astroiden. Laget av Sam Derbyshire - English Wikipedia, Public Domain, [Wikimedia Commons](https://commons.wikimedia.org/w/index.php?curid=4361393)
```

Parametriseringen av kurven i {numref}`astroid` er gitt ved

$$
\vec{r}(t) = \left[\cos^3(t), \sin^3(t)\right], \qquad t \in [0, 2\pi].
$$

Bruk en valgfri metode til å estimere lengden av kurven.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*


In [ ]:
import numpy as np

def x(t):
    return NotImplemented

def y(t):
    return NotImplemented

a = NotImplemented
b = NotImplemented
n = NotImplemented
dt = NotImplemented

lengde = 0 
for i in range(n):
    t = NotImplemented

    dx = NotImplemented # Δx
    dy = NotImplemented # Δy
    
    ds = NotImplemented # Δs

    lengde += NotImplemented

print(f"{lengde = :.2f}")

````{dropdown} Løsningsforslag
```python
import numpy as np

def x(t):
    return np.cos(t)**3

def y(t):
    return np.sin(t)**3

a = 0
b = 2 * np.pi
n = 100_000
dt = (b - a) / n

lengde = 0 
for i in range(n):
    t = a + i * dt

    dx = x(t + dt) - x(t)
    dy = y(t + dt) - y(t)
    
    ds = np.sqrt(dx**2 + dy**2)

    lengde += ds

print(f"{lengde = :.2f}")
```
som gir utskriften
```console
lengde = 6.00
```
````

### Oppgave 8

En *deltoide* er i matematikken en kurve som genererer ved at en sirkel ruller langs innsiden av en annen sirkel. 

```{figure} ./animasjoner/deltoid.gif
:name: deltoide

Animasjon som viser hvordan en deltoide genereres. Den røde kurven er *deltoiden*. Av Sam Derbyshiret ved English Wikipedia, CC BY-SA 3.0, [Wikimedia Commons](https://commons.wikimedia.org/w/index.php?curid=10784089)
```

En deltoide er beskrevet av parametriseringen

$$
\vec{r}(t) = \left[2\cos(t) + \cos(2t), 2\sin(t) - \sin(2t)\right], \qquad t \in [0, 2\pi].
$$

Bruk en valgfri metode til å regne ut lengden av kurven.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

def x(t):
    return NotImplemented

def y(t):
    return NotImplemented

a = NotImplemented
b = NotImplemented
n = NotImplemented
dt = NotImplemented

lengde = 0 
for i in range(n):
    t = NotImplemented

    dx = NotImplemented # Δx
    dy = NotImplemented # Δy
    
    ds = NotImplemented # Δs

    lengde += NotImplemented

print(f"{lengde = :.2f}")

### Oppgave 9

Den internasjonale romstasjonen (ISS) går i bane rundt jorda med en fart på ca. 7.7 km/s. En tilnærming til banen dens er gitt ved kurven som er vist i 

```{figure} ./figurer/trefoil_knot.pdf
:name: trefoil_knot

En tilnærming til den interasjonale romstasjonens bane rundt jorden. Kurven er en *perturbert trekløverknute* (hvis en matematiker leser denne påstanden, *please forgive this transgression*).
```
Banen representer to fulle omløp rundt jorden. 

Parametriseringen til kurven er gitt ved

\begin{align*}
x(t) & = (R + r\cos(3t) + d\sin(t)) \cos(2t), \\
y(t) & = (R + r\cos(3t) + d\sin(t)) \sin(2t), \\
z(t) & = r\sin(3t) + d\cos(t),
\end{align*}

der $t \in [0, 2\pi]$ er en parameter (som *ikke* representerer tid!). Her er $R = 6371$ km jordas radius, $r = 400$ km er gjennomsnittsavstanden fra bakken opp til ISS, og $d = 50$ km er avstanden ISS svinger rundt gjennomsnittshøyden. 

Avgjør omtrent hvor lang tid det tar for ISS å gå en rundt rundt jorda. *Hint: du må vite hvor lang banen er.*

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*    

In [ ]:
import numpy as np 

R = 6371 # Jordens radius i km
r = 400 # Banens radius i km
d = 50 # Fluktuasjon i km

def x(t):
    return NotImplemented

def y(t):
    return NotImplemented

def z(t):
    return NotImplemented

a = NotImplemented # Startverdi for parameteren t
b = NotImplemented # Sluttverdi for parameteren t
n = NotImplemented # Antall linjeelementer
dt = NotImplemented # Δt

lengde = 0
for i in range(n):
    t = NotImplemented # Regn ut parameteren
    dx = NotImplemented # Regn ut Δx
    dy = NotImplemented # Regn ut Δy
    dz = NotImplemented # Regn ut Δz

    ds = NotImplemented # Regn ut Δs

    lengde = NotImplemented

fart = NotImplemented # Farten i km/s

periode = NotImplemented # Tiden i sekunder
periode = NotImplemented # Tiden i minutter
periode /= 2 # Del på 2 fordi det er to omløp

print(f"{periode = } minutter.")

````{dropdown} Løsningsforslag
Først må vi regne ut lengden av banen. Deretter antar vi at vi kan bruke en konstant hastighet for å regne ut tiden det tar å gå ett omløp med $t = L / v$ der $L$ er lengden, $v$ er banefarten og $t$ er perioden. Vi må dele denne verdien på to fordi banen vi ser på representer to omløp rundt jorda.

```python
import numpy as np 

R = 6371 # Jordens radius i km
r = 400 # Banens radius i km
d = 50 # Fluktuasjon i km

def x(t):
    return (R + r * np.cos(3 * t) + d * np.sin(t)) * np.cos(2 * t)

def y(t):
    return (R + r * np.cos(3 * t) + d * np.sin(t)) * np.sin(2 * t)

def z(t):
    return r * np.sin(3 * t) + d * np.cos(t)

a = 0
b = 2 * np.pi
n = 100_000
dt = (b - a) / n

lengde = 0
for i in range(n - 1):
    t = a + i * dt
    dx = x(t + dt) - x(t)
    dy = y(t + dt) - y(t)
    dz = z(t + dt) - z(t)

    ds = np.sqrt(dx**2 + dy**2 + dz**2)

    lengde += ds

fart = 7.7 # km/s

periode = lengde / fart # tiden i sekunder
periode /= 60 # tider i minutter
periode /= 2
print(f"{periode = :.2f} minutter.")
```
som gir utskriften
```console
periode = 87.03 minutter.
```
som er ca. 5 minutter kortere enn den sanne omløpstiden til ISS.
````